src/can.rs

```diff
@@ -0,0 +1,510 @@
+            CanFilterData::IdFilter(_id) => None, // TODO: This sucks, we need more info here to figure out the correct value of fr2
+        }
+    }
+}
+
+impl Can {
+    pub fn can(
+        can: stm32::CAN,
+        rx: gpioa::PA11<AF9>,
+        tx: gpioa::PA12<AF9>,
+        apb1: &mut APB1,
+    ) -> Can {
+        apb1.enr().modify(|_, w| w.canen().enabled());
+        can.mcr.modify(|_, w| w.sleep().clear_bit());
+        can.mcr.modify(|_, w| w.inrq().set_bit());
+
+        // Wait for INAK to confirm we have entered initialization mode
+        while !can.msr.read().inak().bit_is_set() {}
+
+        // TODO: actually calculate baud params
+
+        // Our baud rate calc here is aiming for roughly 4000uS total bit time or about 250kbps
+        // Though we actually allow closer to 5500uS total given the sjw setting
+        // Calculations for timing value from http://www.bittiming.can-wiki.info/#bxCAN
+
+        // Baud rate prescaler defines time quanta
+        // tq = (BRP[9:0]+1) x tPCLK
+        let brp: u16 = 4;
+
+        // Resynchronization jump width: number of quanta segments may be expanded to resync
+        // tRJW = tq x (SJW[1:0] + 1)
+        let sjw = 0;
+
+        // Time seg 2
+        // tBS2 = tq x (TS2[2:0] + 1)
+        let ts2 = 3;
+
+        // Time seg 1
+        // tBS1 = tq x (TS1[3:0] + 1)
+        let ts1 = 10;
+
+        can.btr.modify(|_, w| unsafe {
+            w.brp()
+                .bits(brp)
+                .sjw()
+                .bits(sjw)
+                .ts1()
+                .bits(ts1)
+                .ts2()
+                .bits(ts2)
+            //.lbkm()
+            //.set_bit()
+        });
+
+        // Leave initialization mode by clearing INRQ and switch to normal mode
+        can.mcr.modify(|_, w| w.inrq().clear_bit());
+        // Wait for INAK again to confirm the mode switch
+        while !can.msr.read().inak().bit_is_clear() {}
+
+        Can {
+            can,
+            _rx: rx,
+            _tx: tx,
+        }
+    }
+
+    /// Enable CAN event interrupts for `Event`
+    pub fn listen(&mut self, event: Event) {
+        self.can.ier.modify(|_, w| match event {
+            Event::Fifo0Fmp => w.fmpie0().set_bit(),
+            Event::Fifo0Full => w.ffie0().set_bit(),
+            Event::Fifo0Ovr => w.fovie0().set_bit(),
+            Event::Txe => w.tmeie().set_bit(),
+            _ => unimplemented!(),
+        });
+    }
+
+    /// Split the CAN peripheral into a transmitter and associated FIFOs.
+    pub fn split(self) -> (CanTransmitter, CanFifo, CanFifo) {
+        let fifo0 = CanFifo { idx: 0 };
+        let fifo1 = CanFifo { idx: 1 };
+
+        let transmitter = CanTransmitter {
+            _can: self.can,
+            _rx: self._rx,
+            _tx: self._tx,
+        };
+
+        return (transmitter, fifo0, fifo1);
+    }
+
+    pub fn release(self) -> stm32::CAN {
```

Comment:
```suggestion
    pub fn free(self) -> stm32::CAN {
```

And it should release `gpioa::PA11<AF9>` and    `gpioa::PA12<AF9>` as well.